pallets/evm-assertions/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020-2024 Trust Computing GmbH.
+// This file is part of Litentry.
+//
+// Litentry is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+use frame_system::ensure_signed;
+pub use pallet::*;
+use sp_std::vec::Vec;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::OriginFor;
+	use scale_info::TypeInfo;
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[derive(Encode, Decode, Clone, Default, Debug, PartialEq, Eq, TypeInfo)]
+	pub struct Assertion {
+		byte_code: Vec<u8>,
+		secrets: Vec<Vec<u8>>,
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Assertion identifier, usually H160
+		type AssertionId: Default
+			+ Copy
+			+ PartialEq
+			+ core::fmt::Debug
+			+ parity_scale_codec::FullCodec
+			+ TypeInfo;
+	}
+
+	/// Map for storing assertion smart contract bytecode alongside with additional secrets
+	/// Secrets are encrypted with worker's shielding key
+	#[pallet::storage]
+	#[pallet::getter(fn assertions)]
+	pub type Assertions<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssertionId, Assertion, OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		AssertionCreated { id: T::AssertionId, byte_code: Vec<u8>, secrets: Vec<Vec<u8>> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AssertionExists,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((T::DbWeight::get().write, DispatchClass::Normal, Pays::No))]
+		pub fn create_assertion(
+			origin: OriginFor<T>,
+			id: T::AssertionId,
```

Comment:
Hmm I suggest we allocate AssertionId internally, and reserve some sections for Litentry precompiles

Reply:
But it's using H160 right - that could be misleading and might cause (semantic) abuse 😂 

> Litentry precompiles

I mean those contracts wirtten by Litentry ourselves (e.g. A1 - A20)

Reply:
From the client perspective you dont need to be aware about implementation details. All you need to do is to generate unique `H160` identifier and use it during `create` and `request_vc`. I think it's natural to use `H160` because that's how smart contracts are identified - but we can change to another type depends if it's goind to be generated internally or by client. 

I suggest to treat all assertions equally so there is no need for identifier reservation for assertions currently used by IdHub.

Given that, do you still think it's worth it ?

Reply:
Technically there's so such requirement.

I'm thinking of how it looks like in the future - like when we introduce "reputation system" to find and slash bad contracts, it's easier to identify internal and external contracts just by looking at their identifiers. It's less error-prone (for clients to avoid scams) and easier to document.

Internal contracts don't carry along any particular feature, but they are marked as "written by litentry" so people know the trustness of the logic should be at least as high as Litentry the project team.

I see similar implementation to multi asset management, where most projects do draw a line between "internally defined assets" (assets that are registered by the chain team itself) and "externally defined assets" that can be registered arbitrarily by anyone.

So I'm ok to have H160 as identifier but I think it makes sense (at least I don't see any harm) to have a reserved section for Litentry contracts (both primitive precompiles and concrete logic contracts) - like 65536 should be enough? Wdyt?

Reply:
Having a reputation system for assertions is a nice idea. I would like to draw another way of meeting these requirements.

We could extend parachain's storage with `bool` value, let's name it `verified`. Ideally, assertions could be marked as `verified` upon collective decision. The `verified` mark could be also removed in the same manner, for example after detecting that assertion generates false-positive vcs.

All assertions written by the Litentry team could get `verified` mark from the start. It would also open the possibility to mark other assertions as `verified` after auditing their code, for example by the Litentry team.

We can also extend storage with `owner` address. That would allow for detecting which assertions are added by the Litentry team.